### PR TITLE
Add GRR Hunt Client filters - by label and operating system

### LINKED ISSUE #493: Add Label/OS Client Filters to GRR Hunt collectors
It would be useful to apply client filters such as labels and OS type when running GRR hunts.

### PR DESCRIPTION
**Pull request by sydp**
Fixes #493

### CODE REVIEW

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 1]
```diff
@@ -54,8 +63,52 @@ def _CreateHunt(
...
+        label_rule.rule_type = label_rule.LABEL
+        label_rule.label.label_names.append(client_label)
+
+    if self.client_operating_systems:
+      client_operating_systems = set(
+          os for os in self.client_operating_systems.lower().split(',')
+          if os in ('win', 'osx', 'linux'))
+
+      if len(client_operating_systems) != len(self.client_operating_systems):
+        self.ModuleError('Invalid client OS values.', critical=True)
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:90]: Can we move this check to `SetUp()` ? This makes sure that a valid OS version is passed before anything else executes.

File: data/recipes/grr_hunt_file.json [hunk 2]
```diff
@@ -26,6 +29,9 @@
         ["--verify", "Whether to verify the GRR TLS certificate.", true],
         ["--grr_username", "GRR username", "admin"],
         ["--grr_password", "GRR password", "admin"],
-        ["--max_file_size", "Maximum size of files to collect (in bytes).", 5368709120]
+        ["--max_file_size", "Maximum size of files to collect (in bytes).", 5368709120],
+        ["--match_mode", "Match mode of the client rule set (ANY or ALL)", null],
+        ["--client_operating_systems", "Comma-separated list of client operating systems to filter hosts on.", null],
```
Comment (tomchop) [inline data/recipes/grr_hunt_file.json:34]: It would be good to specify the valid options here.

**Review by tomchop — CHANGES_REQUESTED** [review]

**Comment by tomchop** [comment]
Also please fix the mypy issues.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 1 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py]: Done

File: data/recipes/grr_hunt_file.json (hunk 2 above)
Comment (sydp) [reply data/recipes/grr_hunt_file.json]: Done.

**Comment by sydp** [comment]
> Also please fix the mypy issues.

Done!

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 3]
```diff
@@ -54,8 +97,44 @@ def _CreateHunt(
     """
     runner_args = self.grr_api.types.CreateHuntRunnerArgs()
     runner_args.description = self.reason
+
+    if self.match_mode:
+      if self.match_mode.lower() == 'any':
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:102]: We can `lower` the match mode in `HuntSetup()` and assume it's already formatted as expected here.

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 4]
```diff
@@ -25,20 +25,63 @@ class GRRHunt(grr_base.GRRBaseModule, module.BaseModule):  # pylint: disable=abs
...
+
+    Args:
+      match_mode (Optional[str]): match mode of the client rule set (ALL or
+          ANY).
+      client_operating_systems (Optional[str]): a comma separated list of client
+          OS types (win, osx or linux).
+      client_labels (Optional[str]): a comma separated list of client labels.
+    """
+    if match_mode:
+      if match_mode.upper() not in ('ALL', 'ANY'):
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:61]: Super nit: for consistency i'd use either lower or upper throughout this function, but stick to one.

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 5]
```diff
@@ -54,8 +97,44 @@ def _CreateHunt(
...
     runner_args = self.grr_api.types.CreateHuntRunnerArgs()
     runner_args.description = self.reason
+
+    if self.match_mode:
+      if self.match_mode.lower() == 'any':
+        match_mode = runner_args.client_rule_set.MATCH_ANY
+      elif self.match_mode.lower() == 'all':
+        match_mode = runner_args.client_rule_set.MATCH_ALL
+      else:
+        self.ModuleError(f'Unknown match mode {self.match_mode}', critical=True)
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:107]: You're already making sure that match mode is good in HuntSetup(), no? If so, no need to check here.

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 6]
```diff
@@ -54,8 +97,44 @@ def _CreateHunt(
...
+      for client_label in self.client_labels:
+        label_rule = runner_args.client_rule_set.rules.add()
+
+        label_rule.rule_type = label_rule.LABEL
+        label_rule.label.label_names.append(client_label)
+
+    if self.client_operating_systems:
+      client_operating_systems = set(
+          os for os in self.client_operating_systems
+          if os in ('win', 'osx', 'linux'))
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:121]: Same comment here - I'd say it's safe to assume that `client_operating_systems` has been validated, so no need for th `if` condition here. In fact, you can remove this block completely and just loop over `self.client_operating_systems` in L 123

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 7]
```diff
@@ -25,20 +25,63 @@ class GRRHunt(grr_base.GRRBaseModule, module.BaseModule):  # pylint: disable=abs
...
+      client_labels (Optional[str]): a comma separated list of client labels.
+    """
+    if match_mode:
+      if match_mode.upper() not in ('ALL', 'ANY'):
+        self.ModuleError(f'Unknown match mode {self.match_mode}', critical=True)
+
+      self.match_mode = match_mode
+
+    if client_operating_systems:
+      normalised_client_operating_systems = list(set(
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:67]: It looks like you're using a `set` in `_CreateHunt`, so you can just skip the `list` here?

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 8]
```diff
@@ -100,8 +179,11 @@ def SetUp(self,
             grr_username: str,
             grr_password: str,
             max_file_size: str,
-            approvers: Optional[str]=None,
-            verify: bool=True) -> None:
+            approvers: Optional[str] = None,
+            verify: bool = True,
+            match_mode: Optional[str] = None,
+            client_operating_systems: Optional[str] = None,
+            client_labels: Optional[str] = None) -> None:
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:186]: Actually, none of these arguments should be optional (we always provide the full list of kwargs, and ther default values are defined at recipe level). Some cleanup happened but apparently not entirely, heh. Could you remove the `Optional` checks and default values here?

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 9]
```diff
@@ -153,31 +242,34 @@ class GRRHuntFileCollector(GRRHunt):
...
```
Comment (tomchop) [inline dftimewolf/lib/collectors/grr_hunt.py:272]: Please see my note in the other SetUp function

**Review by tomchop — CHANGES_REQUESTED** [review]

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 3 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py]: Done.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 4 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py]: Switched to lower for consistency.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 5 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py]: Yes that's correct.  Removed the check.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 6 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py]: Removed.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 7 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py:67]: list() is required to keep mypy happy.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 8 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py]: Cleaned up.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 9 above)
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py]: Cleaned up.

File: dftimewolf/lib/collectors/grr_hunt.py (hunk 7 above)
Comment (tomchop) [reply dftimewolf/lib/collectors/grr_hunt.py:67]: This is probably because you type it as a `List[str]` in L44. Have you tried using `Set[str]` instead?

**Review by tomchop — CHANGES_REQUESTED** [review]
One last mypy nit, otherwise this looks good! Thanks a lot :)

**Comment by sydp** [comment]
> One last mypy nit, otherwise this looks good! Thanks a lot :)

Updated!

**Review by tomchop — APPROVED** [review]

**Pull request merged**